ring/batch.go

```diff
@@ -4,13 +4,44 @@ package ring
 
 import (
 	"context"
+	"errors"
 	"fmt"
 	"sync"
 
 	"go.uber.org/atomic"
 	"google.golang.org/grpc/status"
 )
 
+type Error interface {
+	error
+	IsClientError() bool
+}
+
+type batchError struct {
+	isClientError bool
+	err           error
+}
+
+func newError(err error) Error {
+	var e Error
+	if errors.As(err, &e) {
```

Comment:
i'm not sure about this. This will unwrap the error and also forces `e` to escape to the heap. Can we instead do a simple type assertion `if e, ok := err.(Error); ok {`? That will retain the error and it will be easier to explain how the contract works